Get data for standalone mgt-person-card

beth-panx opened the pull request:
Closes #324
Closes #321 
Closes #293 

### PR Type
- Feature

### Description of the changes
You can now use <mgt-person-card person-query="me"></mgt-person-card> to display person-card alone without nested with mgt-person

### PR checklist
- [x] Project builds (`npm run build`) and changes have been tested in supported browsers (including IE11)
- [x] All public classes and methods have been documented
- [x] Added appropriate [documentation](https://github.com/microsoftgraph/microsoft-graph-docs/pull/7610)
- [x] License header has been added to all new source files (`npm run setLicense`)
- [x] Contains **NO** breaking changes

Docs: https://github.com/microsoftgraph/microsoft-graph-docs/pull/7610

shweaver-MSFT commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -443,6 +462,16 @@ export class MgtPersonCard extends MgtTemplatedComponent {
+      return;
+    }
+
+    if (this.personDetails) {
```
Make sure any change in `userId`, `personQuery`, or `inheritDetails` invalidates the current `personDetails`. For reference, check out the `attributeChangedCallback` function of `mgt-person`.

beth-panx replied:
Done

shweaver-MSFT commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -480,22 +480,22 @@ export class MgtPerson extends MgtTemplatedComponent {
+      let person = null;
+      let photo = null;
 
       if (this.userId) {
```
I actually prefer the batch logic, but I do like that you've moved the logic into separate `*.graph.ts` files. 

Perhaps there is a happy middle, with a function that returns the results of the batch request (person and photo) combined. In fact, if you use the `IDynamicPerson` object from `mgt-person.ts`, you can do it something like this in whichever `*.graph.ts`:
1. Make the Batch request for person and photo.
1. Cast the person as an `IDyanimcPerson`
1. Append the photo to the `person.personImage` property.
1. Return the person.

This could work for both `mgt-person` and `mgt-person-card` as well, and you won't have to make multiple Graph calls 😊

beth-panx replied:
Just to clarify - we will have single graph calls to get person, get photo, and a batch call to get both person and photo?

shweaver-MSFT replied:
Yes, assuming all cases are explicitly used... Add whatever methods make most sense for how we use them. The Graph helper functions aren't intended to be exposed.

beth-panx replied:
Makes sense. Easier to read that way too.

beth-panx replied:
done

shweaver-MSFT commented on src/graph/graph.user.ts:
```diff
@@ -15,11 +15,16 @@ import { prepScopes } from '../utils/GraphHelpers';
-    .get();
+export async function getMe(graph: IGraph): Promise<User> {
+  const batch = graph.createBatch();
+  batch.get('user', 'me', ['user.read']);
```
Batch seems unnecessary here. We're only making one get request.

beth-panx replied:
For context originally mgt-person makes a batch call to get 'me' so when I refactored that code out I've updated it here. Can def keep this as single request if we agree to the former question ☝🏼to create a batch to get both person and person photo. I am ok with either implementation.

shweaver-MSFT replied:
Seeing how `mgt-person` and `mgt-person-card` do pretty much the same data calls, you may find we don't need this at all. It's all about whatever works best for the use case.

beth-panx replied:
done

shweaver-MSFT commented on src/components/mgt-person/mgt-person.ts:
```diff
@@ -480,22 +480,22 @@ export class MgtPerson extends MgtTemplatedComponent {
       return;
     }
```
`mgt-person` has a check for personDetails, but no image. Make sure to do something similar in `mgt-person-card`.

beth-panx replied:
Done

beth-panx commented:
ping @shweaver-MSFT @nmetulev 
Updated renderContactIcons to include different IDynamicPerson casted types in order to fix bug #293 where icons of chat and phone do not show up in mgt-people => mgt-person-card
Added logic to handle get more details for mgt-person-card when userPrincipalName is not present or when email is not present. Also included different IDynamicPerson types. This should fix #324
Original implementation should cover #321 where user-id and person-query are added as attributes. I have a branch created for docs. Once this is merged, I will submit the pr on docs repo.

nmetulev commented:
@beth-panx, could you resolve conflict so we can test the PR with the latest changes from #357

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -267,10 +298,13 @@ export class MgtPersonCard extends MgtTemplatedComponent {
     let chat: TemplateResult;
-    if (userPerson.userPrincipalName) {
+
+    if (userPerson.userPrincipalName || personPerson.userPrincipalName) {
```
This is not necessary since it's the same property name (`userPrincipalName`). The compiled js would essentially end up being: `if (person.userPrincipalName || person.userPrincipalName)`

```suggestion
    if (userPerson.userPrincipalName) {
```

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -290,7 +324,11 @@ export class MgtPersonCard extends MgtTemplatedComponent {
     // Phone
     let phone: TemplateResult;
-    if (userPerson.businessPhones && userPerson.businessPhones.length > 0) {
+    if (
+      (userPerson.businessPhones && userPerson.businessPhones.length > 0) ||
+      (personPerson.phones && personPerson.phones.length > 0) ||
+      (contactPerson.businessPhones && contactPerson.businessPhones.length > 0)
+    ) {
```
Same here. Can be simplified:

```suggestion
    if (
      (userPerson.businessPhones && userPerson.businessPhones.length > 0) ||
      (personPerson.phones && personPerson.phones.length > 0))
    ) {
```

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -290,7 +324,11 @@ export class MgtPersonCard extends MgtTemplatedComponent {
 
     // Phone
     let phone: TemplateResult;
```
This is a great improvement, can you also update it below in the `renderAdditionalDetails` function? The phone doesn't show up when expanded otherwise

beth-panx replied:
Did you mean `renderContactDetails?`

nmetulev replied:
Agh, yes, that's right, I meant `renderContactDetails`

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -450,6 +488,39 @@ export class MgtPersonCard extends MgtTemplatedComponent {
    */
   protected async loadState() {
+    const provider = Providers.globalProvider;
+    const graph = provider.graph.forComponent(this);
```
This causes exceptions if `provider` is null. Move this line below the check for signed in user

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -450,6 +488,39 @@ export class MgtPersonCard extends MgtTemplatedComponent {
+      }
+    }
+
     if (this.inheritDetails) {
```
this should be `else if`, otherwise the work above could be overwritten.

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -460,15 +531,28 @@ export class MgtPersonCard extends MgtTemplatedComponent {
-    }
+    } else {
+      // Use userId or 'me' query to get the person and image
+      if (this.userId || this.personQuery === 'me') {
```
there is no shared logic between these two if statements, they can be moved one level up as part of the `else statement.

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -460,15 +531,28 @@ export class MgtPersonCard extends MgtTemplatedComponent {
-    const provider = Providers.globalProvider;
-    if (!provider || provider.state !== ProviderState.SignedIn) {
-      return;
+        if (people && people.length) {
```
This should an `else if` and the above `return` statement should be removed

nmetulev commented on stories/components/personCard.stories.js:
```diff
@@ -21,22 +21,27 @@ export default {
-  <div style="margin:2em 0 0 1em;font-family:segoe ui;color:#323130;font-size:12px">
-    (Hover on person to view Person Card)
-  </div>
+export const personCard = () => html`
```
is this rename going to change the url and break our docs?

beth-panx replied:
Good catch. Maybe we should create aka links and update urls in aka links... Just a thought. Is it difficult to update links in docs in terms of process?

nmetulev replied:
would aka links work in the iframe? re updating docs, we just need to time it right so they get updated at the same time. @elisenyang, is it too late to change the url in the docs?

elisenyang replied:
@beth-panx @nmetulev does this need to be updated for 1.2?

beth-panx replied:
I have added a new one instead of changing this one. So we are good.

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -450,6 +488,39 @@ export class MgtPersonCard extends MgtTemplatedComponent {
+
+    // check if personDetail already populated
+    if (this.personDetails) {
+      // in some cases we might only have name or email, but need to find the image
+      // use @ for the image value to search for an image
+      if (this.personImage === '@' && !this.personDetails.personImage) {
+        this.loadImage();
+      }
+
+      if (this.personDetails.id) {
+        const user = this.personDetails as MicrosoftGraph.User;
+        const person = this.personDetails as MicrosoftGraph.Person;
+        const contact = this.personDetails as MicrosoftGraph.Contact;
+
+        // check if there's userPrincipalName in personDetails
+        if (user.userPrincipalName || person.userPrincipalName) {
+          // check if there's email present in personDetails
+          if (user.mail || person.scoredEmailAddresses || contact.emailAddresses) {
+            return;
+          }
+        } else {
+          this.personDetails = await getUser(graph, this.personDetails.id);
+        }
+      }
```
We should check for `id` or `userPrincipalName`, not just `id` - both could be used interchangeably and both work the same way with the user api.

Also, you can check for the image at the same time using your new `getUserWithPhoto` function

Here is a suggested fix:

```suggestion
      const user = this.personDetails as MicrosoftGraph.User;
      const id = user.userPrincipalName || user.id;

      // if we have an id but no email, we should get data from the graph
      if (id && !getEmailFromGraphEntity(user)) {
        try {
          const person = await getUserWithPhoto(graph, id);

          this.personDetails = person;
          this.personImage = this.getImage();
        } catch {
          // nop
        }
      } else if (this.personImage === '@' && !this.personDetails.personImage) {
        // in some cases we might only have name or email, but need to find the image
        // use @ for the image value to search for an image
        this.loadImage();
      }
```

beth-panx commented:
ping @nmetulev your comments have all been addressed.

nmetulev commented on src/components/mgt-person-card/mgt-person-card.ts:
```diff
@@ -460,15 +528,23 @@ export class MgtPersonCard extends MgtTemplatedComponent {
+      if (people && people.length) {
+        this.personDetails = people[0];
+
+        if (this.personImage === '@') {
+          this.loadImage();
+        }
```
I think this is causing the issue with the image not showing when person-query is used. Should probably be:

```suggestion
          this.loadImage();
```

Could you make the same fix for person?

beth-panx replied:
On it

beth-panx replied:
Done

beth-panx commented:
> Other than the fix for person image not showing up when using person-query, and the rename of the story, it looks good.
> 
> Make sure to also create a documentation PR for the new properties.

Docs pr is submitted. Not sure if I added correct tags. Let me know.

nmetulev commented:
Awesome, I updated the description to include the docs. I think we are good to go here
